### PR TITLE
Mixed Dimension Mesh Support for FEMContext

### PR DESCRIPTION
**Pull request by pbauman**
Updated `FEMContext` to support cases of mixed dimension meshes. This is accomplished by caching `FEAbstract*` etc. now for all dimensions present in the mesh as well as `FEType`. It is fully backward compatible with dimension dependent calls defaulting to element dimension if it's guaranteed (or at least supposed to be) available or to mesh dimension if it's not. `FEMSystem` examples ran for me and I have a working GRINS test case that I'll create a PR for in a bit, but at least wanted a chance for eyes to be put on this.

### CODE REVIEW

File: include/systems/fem_context.h [hunk 1]
```diff
@@ -776,18 +842,22 @@ class FEMContext : public DiffContext
...
   /**
    * Pointers to the same finite element objects, but indexed
-   * by variable number
+   * by variable number. We store FE objects for each element dimension
+   * present in the mesh, except for edge_fe_var which only applies
+   * for 3D elements.
    */
-  std::vector<FEAbstract*> _element_fe_var;
-  std::vector<FEAbstract*> _side_fe_var;
+  std::map<unsigned char, std::vector<FEAbstract*> > _element_fe_var;
```
Comment (roystgnr) [inline include/systems/fem_context.h]: Why a map here?  The extra overhead from a single map node is probably as much as the extra overhead of a size 3 vector would be.

File: include/systems/fem_context.h [hunk 2]
```diff
@@ -776,18 +842,22 @@ class FEMContext : public DiffContext
...
+  std::map<unsigned char, std::vector<FEAbstract*> > _side_fe_var;
   std::vector<FEAbstract*> _edge_fe_var;
```
Comment (roystgnr) [inline include/systems/fem_context.h]: No dimensionality here because "edge" only makes sense if you're on a 3D element?

Comment (pbauman) [inline include/systems/fem_context.h]: Correct. I thought I had a comment to that effect... I'll add one.

File: include/systems/fem_context.h (hunk 1 above)
Comment (pbauman) [inline include/systems/fem_context.h]: Laziness. Made asserts easier to code later. I'll update to vector.

File: src/systems/fem_context.C [hunk 3]
```diff
@@ -64,78 +66,98 @@ FEMContext::FEMContext (const System &sys)
...
-  for (std::map<FEType, FEAbstract *>::iterator i = _side_fe.begin();
-       i != _side_fe.end(); ++i)
-    delete i->second;
-  _side_fe.clear();
+  for (std::map<unsigned char, std::map<FEType, FEAbstract *> >::iterator d = _element_fe.begin();
+       d != _element_fe.end(); ++d)
+    for (std::map<FEType, FEAbstract *>::iterator i = (d->second).begin();
+         i != (d->second).end(); ++i)
+      delete i->second;
+
```
Comment (roystgnr) [inline src/systems/fem_context.C]: This looks correct, but is scary.  Could you churn a GRINS run through valgrind (after switching map to vector if you take my advice about that) and make sure we're not leaking or mis-deleting anything here?

Comment (pbauman) [inline src/systems/fem_context.C]: Will do.

**Comment by roystgnr** [comment]
Nothing obviously wrong here; I like it.

File: include/systems/fem_context.h (hunk 2 above)
Comment (pbauman) [inline include/systems/fem_context.h]: OK, there actually is a comment there, it's just not showing up on the diff because the diff is ugly.

**Comment by pbauman** [comment]
OK, updated to using vector data structure instead. GRINS make check runs and my test case runs. Will run through valgrind tomorrow (I don't trust valgrind on the Mac).

**Comment by pbauman** [comment]
Ran through valgrind this morning and all is well. Will merge later today if no further suggestions and/or complaints.

**Comment by roystgnr** [comment]
:+1:

**Pull request merged**